File: bindings/ruby/lib/cconfigspace/expression_parser.rb
def silence_warnings(&block)
  warn_level = $VERBOSE
  $VERBOSE = nil
  result = block.call
  $VERBOSE = warn_level
  result
end

silence_warnings {
  require 'whittle'
}

undef silence_warnings
module CCS
  AssociativityMap = {
    :CCS_ASSOCIATIVITY_TYPE_LEFT_TO_RIGHT => :left,
    :CCS_ASSOCIATIVITY_TYPE_RIGHT_TO_LEFT => :right
  }
  class ExpressionParser < Whittle::Parser
    class << self
      attr_accessor :context
      attr_accessor :bind
      attr_accessor :mutex
      attr_reader :parser
    end

    @mutex = Mutex.new

    def parse(input, context: nil, binding: TOPLEVEL_BINDING, **options)
      self.class.mutex.synchronize {
        self.class.context = context
        self.class.bind = binding
        expr = super(input, options)
        self.class.context = nil
        self.class.bind = nil
        expr
      }
    end

    ExpressionSymbols.reverse_each { |k, v|
      next unless v
      next if k == :CCS_EXPRESSION_TYPE_POSITIVE || k == :CCS_EXPRESSION_TYPE_NEGATIVE
      associativity = AssociativityMap[ExpressionAssociativity[k]]
      precedence = ExpressionPrecedence[k]
      rule(v) % associativity ^ precedence
    }

    rule(:wsp => /\s+/).skip!
    rule("(")
    rule(")")
    rule("[")
    rule("]")
    rule(",")

    rule(:none => Regexp.new(TerminalRegexp[:CCS_TERMINAL_TYPE_NONE])).as { |b|
      Expression::Literal::new(value: nil) }
    rule(:true => Regexp.new(TerminalRegexp[:CCS_TERMINAL_TYPE_TRUE])).as { |b|
      Expression::Literal::new(value: true) }
    rule(:false => Regexp.new(TerminalRegexp[:CCS_TERMINAL_TYPE_FALSE])).as { |b|
      Expression::Literal::new(value: false) }
    rule(:float => Regexp.new(TerminalRegexp[:CCS_TERMINAL_TYPE_FLOAT])).as {|num|
      Expression::Literal::new(value: Float(num)) }
    rule(:integer => Regexp.new(TerminalRegexp[:CCS_TERMINAL_TYPE_INTEGER])).as { |num|
      Expression::Literal::new(value: Integer(num)) }
    rule(:identifier => /[:a-zA-Z_][a-zA-Z_0-9]*/)
    rule(:variable) do |r|
      r[:identifier].as { |identifier| Expression::Variable::new(parameter: context.kind_of?(Context) ? context.parameter_by_name(identifier) : context[identifier]) }
    end
    rule(:string => Regexp.new(TerminalRegexp[:CCS_TERMINAL_TYPE_STRING])).as { |str|
      Expression::Literal::new(value: eval(str)) }


    rule(:value) do |r|
      r[:none]
      r[:true]
      r[:false]
      r[:string]
      r[:user_defined]
      r[:variable]
      r[:float]
      r[:integer]
    end

    rule(:list_item) do |r|
      r[:list_item, ",", :value].as { |l, _, e| l.push e }
      r[:value].as { |e| [e] }
    end

    rule(:list) do |r|
      r["[", :list_item, "]"].as { |_, l, _| Expression::List::new(values: l) }
      r["[", "]"].as { |_, _| Expression::List::new(values: []) }
    end

    rule(:user_defined) do |r|
      r[:identifier, "(", :list_item, ")"].as do |e, _, l, _|
        m = bind.receiver.method(e.to_sym)
        evaluate = lambda { |expr, *args| m.call(*args) }
        Expression::UserDefined.new(name: e, nodes: l, eval: evaluate)
      end
      r[:identifier, "(", ")"].as do |e, _, _|
        m = bind.receiver.method(e.to_sym)
        evaluate = lambda { |expr, *args| m.call }
        Expression::UserDefined.new(name: e, eval: evaluate)
      end
    end

    rule(:expr) do |r|
      r["(", :expr, ")"].as { |_, exp, _| exp }
      ExpressionSymbols.reverse_each { |k, v|
        next unless v
        next if v == "#"
        arity = ExpressionArity[k]
        if arity == 1
          r[v, :expr].as { |_, a| Expression.expression_map[k].new(node: a) }
        else
          r[:expr, v, :expr].as { |a, _, b| Expression.expression_map[k].new(left: a, right: b) }
        end
      }
      r[:expr, "#", :list].as { |v, _, l| Expression::In.new(left: v, right: l) }
      r[:value]
    end

    start(:expr)

  end

  @parser = ExpressionParser.new

  def self.parse(input, context: nil, binding: TOPLEVEL_BINDING, **params)
    @parser.parse(input, context: context, binding: binding, **params)
  end

end

File: bindings/python/cconfigspace/expression_parser.py
from parglare import Parser, Grammar
from .expression import Expression, ExpressionType, AssociativityType, ccs_expression_precedence, ccs_expression_associativity, ccs_expression_symbols, ccs_expression_arity, TerminalType, ccs_terminal_precedence, ccs_terminal_regexp

_associativity_map = {
  AssociativityType.LEFT_TO_RIGHT: "left",
  AssociativityType.RIGHT_TO_LEFT: "right"
}
ccs_grammar = "expr: '(' expr ')'\n"

for i in range(ExpressionType.OR, ExpressionType.IN):
  if ccs_expression_arity[i] == 1:
    ccs_grammar += "    | '{}' expr {{{}, {}}}\n".format(ccs_expression_symbols[i], _associativity_map[ccs_expression_associativity[i].value], ccs_expression_precedence[i])
  else:
    ccs_grammar += "    | expr '{}' expr {{{}, {}}}\n".format(ccs_expression_symbols[i], _associativity_map[ccs_expression_associativity[i].value], ccs_expression_precedence[i])

ccs_grammar += "    | expr '{}' list {{{}, {}}}".format(ccs_expression_symbols[ExpressionType.IN], _associativity_map[ccs_expression_associativity[ExpressionType.IN].value], ccs_expression_precedence[ExpressionType.IN])
ccs_grammar += r"""
    | value;
list: '[' list_item ']'
    | '[' ']';
user_defined: identifier '(' list_item ')'
            | identifier '(' ')';
list_item: list_item ',' value
         | value;
value: none
     | true
     | false
     | string
     | user_defined
     | variable
     | integer
     | float;

variable: identifier;

terminals
none: /%s/ {%d};
true: /%s/ {%d};
false: /%s/ {%d};
identifier: /%s/ {%d};
string: /%s/ {%d};
integer: /%s/ {%d};
float: /%s/ {%d};
""" % (ccs_terminal_regexp[TerminalType.NONE], ccs_terminal_precedence[TerminalType.NONE],
       ccs_terminal_regexp[TerminalType.TRUE], ccs_terminal_precedence[TerminalType.TRUE],
       ccs_terminal_regexp[TerminalType.FALSE], ccs_terminal_precedence[TerminalType.FALSE],
       ccs_terminal_regexp[TerminalType.IDENTIFIER], ccs_terminal_precedence[TerminalType.IDENTIFIER],
       ccs_terminal_regexp[TerminalType.STRING], ccs_terminal_precedence[TerminalType.STRING],
       ccs_terminal_regexp[TerminalType.INTEGER], ccs_terminal_precedence[TerminalType.INTEGER],
       ccs_terminal_regexp[TerminalType.FLOAT], ccs_terminal_precedence[TerminalType.FLOAT])
_actions = {}
_expr_actions = [ lambda _, n: n[1] ]

for i in range(ExpressionType.OR, ExpressionType.LIST):
  if ccs_expression_arity[i] == 1:
    _expr_actions.append((lambda s: lambda _, n: Expression.EXPRESSION_MAP[s](node = n[1]))(i))
  else:
    _expr_actions.append((lambda s: lambda _, n: Expression.EXPRESSION_MAP[s](left = n[0], right = n[2]))(i))
_expr_actions.append(lambda _, n: n[0])
_actions["expr"] = _expr_actions
_actions["list"] = [
  lambda _, n: Expression.List(values = n[1]),
  lambda _, n: Expression.List(values = [])
]

def wrap_user_function(proc):
  def wrap(expr, *args):
    return proc(*args)
  return wrap

_actions["user_defined"] = [
  lambda p, n: Expression.UserDefined(name = n[0], nodes = n[2], evaluate = wrap_user_function(p.extra[1][n[0]])),
  lambda p, n: Expression.UserDefined(name = n[0], evaluate = wrap_user_function(p.extra[1][n[0]]))
]
_actions["list_item"] = [
  lambda _, n: n[0] + [n[2]],
  lambda _, n: [n[0]]
]
_actions["variable"] = [
  lambda p, n: Expression.Variable(parameter = p.extra[0][n[0]] if isinstance(p.extra[0], dict) else p.extra[0].parameter_by_name(n[0]))
]
_actions["none"] = lambda _, value: Expression.Literal(value = None)
_actions["true"] = lambda _, value: Expression.Literal(value = True)
_actions["false"] = lambda _, value: Expression.Literal(value = False)
_actions["identifier"] = lambda _, value: value
_actions["string"] = lambda _, value: Expression.Literal(value = eval(value))
_actions["float"] = lambda _, value: Expression.Literal(value = float(value))
_actions["integer"] = lambda _, value: Expression.Literal(value = int(value))

_g = Grammar.from_string(ccs_grammar)

parser = Parser(_g, actions=_actions)

def parse(expr, context = {}, binding = {}):
  return parser.parse(expr, extra = (context, binding))
